New lint: `manual_is_multiple_of`

samueltardieu opened the pull request:
I've added a `min_divisor` configuration option, default to 4, to not trigger on smaller divisibility operations. I would prefer not to lint `if a & 1 == 0` as `if a.is_multiple_of(2)` by default because the former is very idiomatic in systems (and embedded) programming.

This PR will stay in draft mode until the next rustup which will mark `unsigned_is_multiple_of` stable for Rust 1.87.0.

What should its category be? I've used "complexity", but "pedantic" might be suitable as well.

Close #14289
changelog: [`manual_is_multiple_of`]: new lint
r? ghost

joshtriplett commented:
Having a `min_divisor` configuration option seems reasonable, but I think that option should default to flagging every instance of this. Large codebases that want to keep the `& 1 == 0` pattern can set that option.

samueltardieu commented:
> Having a `min_divisor` configuration option seems reasonable, but I think that option should default to flagging every instance of this. Large codebases that want to keep the `& 1 == 0` pattern can set that option.

I've added an extra commit to test this with lintcheck. There was only one hit in Clippy sources (except for tests).

samueltardieu commented:
@joshtriplett The [lintcheck output](https://github.com/rust-lang/rust-clippy/actions/runs/13527240339?pr=14292) seems quite reasonable indeed.

y21 commented:
I agree with linting the `a % b == 0` pattern, but I'm unsure about the `&` one... Something like `(x & 0b111) == 0` feels like a very general expression.
I'm a bit concerned that linting here has the opposite effect: emitting warnings on code that has a totally different semantic meaning where `x.is_multiple_of(8)` would be a very obfuscated way of checking that the last 3 bits aren't set.

joshtriplett commented:
First of all, please don't treat that as a blocker; it's not as important as the `% N == 0` case.

That said: I've seen many, many codebases which use `& (PAGE_SIZE - 1) == 0` and similar instead of `% PAGE_SIZE == 0`, when they know the value is a power of two.

We may need to tune the heuristic to make sure it doesn't produce false positives, and for instance we probably *shouldn't* do it for `& 1 == 0` because that could be a test of a flag, but I think things like `& 4095 == 0` or `& (4096 - 1) == 0` or `& (CONST_EQUAL_TO_4096 - 1) == 0` are pretty clear.

samueltardieu commented:
What about applying the min-divisor option (maybe using another name, like min-one-bits to represent the number of ones) only to the `&(2^n-1)` form of the lint, with a sensible default such as 5 bits? The `%` form would be linted unconditionally.

joshtriplett commented:
@samueltardieu Sounds reasonable. I would say the default should flag 7 and 15 by default, and maybe 3, but not 1.

samueltardieu commented:
Done, see the [toplevel comment](https://github.com/rust-lang/rust-clippy/pull/14292#issue-2878239415) for the updated option.

samueltardieu commented:
rustup happened, PR ready.
r? clippy

samueltardieu commented:
Rebased.

samueltardieu commented:
Rebased

blyxyas commented on clippy_lints/src/if_not_else.rs:
```diff
@@ -48,13 +48,6 @@ declare_clippy_lint! {
 
 declare_lint_pass!(IfNotElse => [IF_NOT_ELSE]);
 
-fn is_zero_const(expr: &Expr<'_>, cx: &LateContext<'_>) -> bool {
```
I think that we can remove this, as we already have the same function in utils.

samueltardieu replied:
> I think that we can remove this, as we already have the same function in utils.

I'm not sure I follow: your comment seems to precisely point to code that does this removal.

blyxyas replied:
Oh, sorry. My PR reviewing UI in neovim was highlighting these two diffs as green, so I thought you added it. Sorry for the confusion.

samueltardieu replied:
No problem, I wondered for a few seconds if I had done something weird while juggling with my patch stacks in jj.

blyxyas commented on clippy_lints/src/operators/manual_is_multiple_of.rs:
```diff
@@ -0,0 +1,115 @@
+        return None;
+    };
+
+    matches!(cx.typeck_results().expr_ty_adjusted(operand).kind(), ty::Uint(_)).then_some(operand)
```
Are we interested in also accounting for constants?

samueltardieu replied:
What do you mean? `CONST & MASK == 0` will already be checked if in an unsigned context. Are you suggesting we lint them if both are constants and typed as signed values?

samueltardieu replied:
(gentle ping to @blyxyas)

blyxyas commented on clippy_lints/src/operators/mod.rs:
```diff
@@ -943,6 +983,7 @@ impl<'tcx> LateLintPass<'tcx> for Operators {
                     rhs,
                     self.modulo_arithmetic_allow_comparison_to_zero,
                 );
+                manual_is_multiple_of::check(cx, e, op.node, lhs, rhs, self.min_and_mask_size, self.msrv);
```
We probably want to move it into `e.span.from_expansion()`, I'm not sure why there are so many lints operating outside of that range, if Clippy usually doesn't want to work in external macros.

samueltardieu replied:
Yes.

samueltardieu replied:
I've also added a test to ensure we don't lint in external macros.

samueltardieu commented on clippy_lints/src/operators/identity_op.rs:
```diff
@@ -170,9 +170,7 @@ fn is_allowed(cx: &LateContext<'_>, cmp: BinOpKind, left: &Expr<'_>, right: &Exp
         // `1 << 0` is a common pattern in bit manipulation code
-        && !(cmp == BinOpKind::Shl
-            && ConstEvalCtxt::new(cx).eval_simple(right) == Some(Constant::Int(0))
-            && ConstEvalCtxt::new(cx).eval_simple(left) == Some(Constant::Int(1)))
```
Here also, this comment points to code which already does exactly what you are suggesting.

samueltardieu commented on clippy_utils/src/consts.rs:
```diff
@@ -958,3 +958,17 @@ fn field_of_struct<'tcx>(
+}
+
+/// Check if `expr` evaluates to an integer constant of 0.
+pub fn is_zero_integer_const(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
```
We don't have many `#[inline]` in `clippy_utils`, but those correspond to simple HIR node matchings, so I guess this would be ok to add them here as well. I'll add them to both `integer_const` and `is_zero_integer_const` functions.

samueltardieu commented on clippy_lints/src/operators/manual_is_multiple_of.rs:
```diff
@@ -0,0 +1,115 @@
+) {
+    if let Some(operand) = uint_compare_to_zero(cx, op, lhs, rhs)
+        && let ExprKind::Binary(lhs_op, lhs_left, lhs_right) = operand.kind
+        && msrv.meets(cx, msrvs::UNSIGNED_IS_MULTIPLE_OF)
```
Good point.

samueltardieu commented:
@blyxyas Should I reassign?

blyxyas commented:
Oh, I got here I review I never submitted ;(

samueltardieu commented on clippy_lints/src/operators/manual_is_multiple_of.rs:
```diff
@@ -0,0 +1,115 @@
+}
+
+/// If `expr` is made of all ones, return the representation of `expr+1` if it is no smaller than
+/// `min_divisor`.
```
I've used the more complete:

```rust
/// If `expr` is provably made of all ones, return the representation of `expr+1` if it is no smaller than
/// `min_divisor`. This will catch expressions of the following forms:
/// - `(1 << A) - 1` where `A` is a constant
/// - integer literals — if it uses hexadecimal, the return value will as well
///
/// The function will not attempt to evaluate non-literal constant expressions, as those may depend on
/// conditional compilation.
```

blyxyas commented:
I've opened the FCP
https://rust-lang.zulipchat.com/#narrow/channel/257328-clippy/topic/FCP.20manual_is_multiple_of

samueltardieu commented:
Removed the `&` part as well as the `min-and-mask-size` option, as discussed in the FCP.

Also updated the clippy version to 1.88.

@rustbot label +S-final-comment-period

samueltardieu commented:
Rebased to fix UI test error in `iter_kv_map.rs`.

samueltardieu commented:
Rebased

samueltardieu commented:
Rebased and set Clippy version to 1.89.

samueltardieu commented:
Rebased

blyxyas reviewed (approved):
LGTM, thanks! :heart:

@flip1995 Does [this message](https://rust-lang.zulipchat.com/#narrow/channel/257328-clippy/topic/FCP.20manual_is_multiple_of/near/515006909) interpret as "I'd approve this if it didn't have binops" or as "I want binops removed from this lint, but don't approve it yet".

Binary operations are now removed, and I think it looks all good for approving (I need the second +1)

samueltardieu commented:
Apart from pinging @blyxyas and @flip1995 in this case, no.

blyxyas reviewed (approved):
LGTM, thanks! :heart:
